src/main/java/com/ontotext/kafka/processor/SinkRecordsProcessor.java

```diff
@@ -0,0 +1,282 @@
+					flushUpdates();
+				}
+			} catch (InterruptedException e) {
+				LOG.info("Thread was interrupted. Shutting down processor");
+				break;
+			}
+		}
+		Thread.interrupted();
+		shutdown();
+	}
+
+	private void shutdown() {
+		// commit any records left before shutdown.
+		LOG.info("Commiting any records left before shutdown");
+		Collection<SinkRecord> records;
+		while ((records = sinkRecords.poll()) != null) {
+			consumeRecords(records);
+		}
+		// final flush after all messages have been batched
+		flushUpdates();
+		if (repository.isInitialized()) {
+			repository.shutDown();
+		}
+	}
+
+	private void consumeRecords(Collection<SinkRecord> messages) {
+		for (SinkRecord message : messages) {
+			recordsBatch.add(message);
+			if (batchSize <= recordsBatch.size()) {
+				flushUpdates();
+			}
+		}
+	}
+
+	private void flushUpdates() {
+		if (!recordsBatch.isEmpty()) {
+			try {
+				retryOperator.execute(this::doFlush, Stage.KAFKA_CONSUME, getClass());
+				if (retryOperator.failed()) {
+					LOG.error("Failed to flush batch updates. This error is due to either repository connection failure, or transaction failure");
+					if (!retryOperator.withinToleranceLimits()) {
+						throw new ConnectException("Error tolerance exceeded.");
+					}
+				}
+			} catch (ConnectException e) {
+				// TODO We would need to push back all records to kafka
+				throw new RuntimeException("Got exception while flushing updates, cannot recover", e);
+			} catch (Exception e) {
+				throw new RuntimeException("Unhandled exception thrown during batch record flush", e);
+			}
+
+		}
+	}
+
+	public Void doFlush() throws RetriableException {
+		long start = System.currentTimeMillis();
+		try (RepositoryConnection connection = repository.getConnection()) {
+			connection.begin();
+			int recordsInCurrentBatch = recordsBatch.size();
+			LOG.trace("Transaction started, batch size: {} , records in current batch: {}", batchSize, recordsInCurrentBatch);
+			while (recordsBatch.peek() != null) {
+				SinkRecord record = recordsBatch.peek();
+				retryOperator.execute(() -> handleRecord(record, connection), Stage.KAFKA_CONSUME, getClass());
+				if (retryOperator.failed()) {
+					LOG.warn("Failed to commit record. Will handle failure, and remove from the batch");
+					errorHandler.handleFailingRecord(record, retryOperator.error());
+					if (!retryOperator.withinToleranceLimits()) {
+						throw new ConnectException("Error tolerance exceeded.");
+					}
+				}
+				recordsBatch.poll();
+
+			}
+			connection.commit();
+			LOG.trace("Transaction commited, Batch size: {} , Records in current batch: {}", batchSize, recordsInCurrentBatch);
+			if (LOG.isTraceEnabled()) {
+				long finish = System.currentTimeMillis();
+				LOG.trace("Finished batch processing for {} ms", finish - start);
+			}
+			return null;
+		} catch (RepositoryException e) {
+			throw new RetriableException(e);
+		}
+	}
+
+	Void handleRecord(SinkRecord record, RepositoryConnection connection) {
```

Comment:
`handleRecord` is passed to `retryOperator.execute()`, which accepts an `Operation<V>`, which is essentially a `Callable<V>`. Therefore `handleRecord` must have a return type. Returning a simple `Object()` does not make much sense here, so we can simply use `Void` as return type and return `null`. 

In case we need something to be returned by this operation, we can change the return type accordingly, but at the moment there is no need.